### PR TITLE
Improve dora load job

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

1. Introduced an iterable UFS status listing interface
2. Avoid calling getStatus from dora worker. Passing the UfsStatus via grpc request  
3. Add integration test for Dora load command
4. Provide users with an option to only load metadata 

### Why are the changes needed?

Currently during load, for each file, a getStatus() is called and this is inefficient. This PR does the listing on the scheduler side and send it to workers via rpc.

### Does this PR introduce any user facing changes?

load command can run with an --metadataOnly option to only load metadata

### CODE REVIEW

File: dora/core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java [hunk 1]
```diff
@@ -282,7 +283,7 @@ public void start() {
           ThreadFactoryUtils.build("scheduler", false));
       mSchedulerExecutor.scheduleAtFixedRate(mWorkerInfoHub::updateWorkers, 0,
           WORKER_UPDATE_INTERVAL, TimeUnit.MILLISECONDS);
-      mSchedulerExecutor.scheduleWithFixedDelay(this::processJobs, 10 * 60 * 1000, 2000,
+      mSchedulerExecutor.scheduleWithFixedDelay(this::processJobs, 10 * 1000, 2000,
```
Comment (elega) [inline dora/core/server/master/src/main/java/alluxio/master/scheduler/Scheduler.java:286]: updated for testing

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -560,38 +562,53 @@ public BlockReader createFileReader(String fileId, long offset, boolean position
...
-        try {
-          if (options.hasUser()) {
-            AuthenticatedClientUser.set(options.getUser());
+    List<LoadFileFailure> errors = Collections.synchronizedList(new ArrayList<>());
+    for (UfsStatus status : ufsStatuses) {
+      String ufsFullPath = status.getUfsFullPath().toString();
+      DoraMeta.FileStatus fs = buildFileStatusFromUfsStatus(status, ufsFullPath);
+      mUfsStatusCache.put(ufsFullPath, fs);
+      if (mMetaStore != null) {
+        mMetaStore.putDoraMeta(ufsFullPath, fs);
```
Comment (elega) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:574]: working on a PR to extract the dora metadata interface 
https://github.com/Alluxio/alluxio/pull/17458#pullrequestreview-1450108606

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 3]
```diff
@@ -147,74 +145,27 @@ public DoraLoadJob(
...
-    FileSystem fs = FileSystem.Factory.create(Scheduler.getInstance().getFileSystemContext());
-    ListStatusPOptions listOptions = ListStatusPOptions
-        .newBuilder()
-        .setRecursive(true).build();
-    mFs = fs;
-    mFileIterator = Optional.of(new FileListFetcher(fs, mPath, listOptions, false));
-  }
+    String ufsRoot = Configuration.getString(PropertyKey.DORA_CLIENT_UFS_ROOT);
+    AlluxioURI ufsRootUri = new AlluxioURI(ufsRoot);
+    mUfs = UnderFileSystem.Factory.create(
```
Comment (elega) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:150]: @lucyge2022 please review this approach

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 4]
```diff
@@ -147,74 +145,27 @@ public DoraLoadJob(
...
+        ufsRoot,
+        UnderFileSystemConfiguration.defaults(Configuration.global()));
 
-  private static class FileListFetcher implements Iterator<URIStatus> {
```
Comment (elega) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:158]: removed

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 5]
```diff
@@ -109,9 +108,8 @@ public class DoraLoadJob extends AbstractJob<DoraLoadJob.DoraLoadTask> {
   private boolean mVerificationEnabled;
 
   // Job states
-  private final LinkedList<String> mRetryFiles = new LinkedList<>();
+  private final Queue<UfsStatus> mRetryFiles = new ArrayDeque<>();
```
Comment (elega) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:111]: i'm not convinced that linked list should be used here

Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:111]: also store the ufs status itself instead of the path string so that we can avoid calling getStatus() again upon failure

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 6]
```diff
@@ -497,24 +443,35 @@ public long getDurationInSec() {
...
-            doraLoadTask.getMyRunningWorker(), response.getFilesList()));
-        for (FileFailure fileFailure : response.getFilesList()) {
-          totalLoadedBytes -= fileFailure.getFile().getLength();
-          if (!isHealthy() || !fileFailure.getRetryable() || !addFilesToRetry(
-              fileFailure.getFile().getAlluxioPath())) {
-            addFileFailure(fileFailure.getFile().getAlluxioPath(), fileFailure.getMessage(),
-                fileFailure.getCode());
+            doraLoadTask.getMyRunningWorker(), response.getFailuresList()));
+        if (response.getFailuresList().size() > 0) {
+          Map<String, LoadFileFailure> failureUfsPath = new HashMap<>();
```
Comment (elega) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:455]: put the corresponding ufs status to the retry list, instead of a path string

File: common/transport/src/main/proto/grpc/block_worker.proto [hunk 7]
```diff
@@ -232,6 +252,16 @@ message BlockStatus {
   optional bool retryable = 4;
 }
 
+message LoadFileFailure {
+  optional string alluxio_path = 1;
```
Comment (elega) [inline common/transport/src/main/proto/grpc/block_worker.proto:256]: this is actually not used... should we keep it?

Comment (jja725) [reply common/transport/src/main/proto/grpc/block_worker.proto:256]: we can remove it if not used

File: dora/core/common/src/main/java/alluxio/underfs/UfsStatus.java [hunk 8]
```diff
@@ -30,6 +33,16 @@ public abstract class UfsStatus {
...
 
+  public AlluxioURI getUfsFullPath() {
+    return mUfsFullPath;
+  }
+
+  public void setUfsFullPath(AlluxioURI ufsFullPath) {
+    mUfsFullPath = ufsFullPath;
+  }
+
+  protected AlluxioURI mUfsFullPath;
```
Comment (lucyge2022) [inline dora/core/common/src/main/java/alluxio/underfs/UfsStatus.java:44]: is mName not ufs path ? is it just the file name?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 3 above)
Comment (lucyge2022) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:150]: think this is better than ufsmanager as it doesnt keep in cache like ufsmanager does @jja725 
but make sure u close the mUfs on exception / job exiting

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 9]
```diff
@@ -147,74 +145,27 @@ public DoraLoadJob(
...
-        if (advance() <= 0) {
-          break;
-        }
-      }
-      URIStatus uriStatus = mFiles.poll();
-      if (uriStatus == null) {
-        throw new NoSuchElementException();
+    try {
+      mUfsStatusIterator = mUfs.listStatusIterable(
+          path, ListOptions.defaults().setRecursive(true), null, 0);
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:158]: plans to support partial listing?

Comment (lucyge2022) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:158]: is this path supposed to be alluxio path or ufs path?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 10]
```diff
@@ -257,13 +200,14 @@ public List<DoraLoadTask> prepareNextTasks() {
...
 
     Map<WorkerInfo, List<DoraLoadTask>> workerToTaskMap = new HashMap<>();
-    for (URIStatus uriStatus : batchBuilder.build()) {
+    for (UfsStatus ufsStatus : batchBuilder.build()) {
       // NOTE: active workers may not reflect all workers at start up,
-      // but hashbased policy will deterministicly only among current recognized active workers
-      WorkerInfo pickedWorker = mWorkerAssignPolicy.pickAWorker(uriStatus.getPath(),
+      // but hash based policy will deterministically only among current recognized active workers
+      WorkerInfo pickedWorker = mWorkerAssignPolicy.pickAWorker(
+          ufsStatus.getUfsFullPath().toString(),
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:207]: dora client actually uses alluxio path to hash to workers instead of ufs path, plz double confirm with others

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 2 above)
Comment (lucyge2022) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:574]: shouldnt we put into cache + metastore after successful load? coz these ufs status comes from scheduler and might be a retried one and might be stale

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 11]
```diff
@@ -225,28 +176,20 @@ public URIStatus next() {
...
-        batchBuilder.add(uriStatus);
-        ++filesToLoad;
-      } catch (AlluxioException | IOException ex) {
-        if (!(ex instanceof FileDoesNotExistException)) {
-          mRetryFiles.offer(path);
-        }
-      }
+        && i++ < startRetryListSize && mRetryFiles.peek() != null) {
+      UfsStatus ufsStatus = mRetryFiles.poll();
+      batchBuilder.add(ufsStatus);
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:186]: when we get the ufsstatus from retry shall we re-fetch the status from ufs to make sure its not stale ?

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 12]
```diff
@@ -560,38 +562,52 @@ public BlockReader createFileReader(String fileId, long offset, boolean position
...
-      futures.add(loadFuture);
+        }, GrpcExecutors.BLOCK_READER_EXECUTOR);
+        futures.add(loadFuture);
+      }
     }
     return Futures.whenAllComplete(futures).call(() -> errors, GrpcExecutors.BLOCK_READER_EXECUTOR);
   }
 
-  private void load(String ufsPath, long mountId, long length) {
+  private void load(String ufsPath, long mountId, long length, boolean reloadMetadata) {
```
Comment (lucyge2022) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:598]: where r u planning to call this func with reloadMetadata = true?

Comment (lucyge2022) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:598]: actually instead of avoid load metadata here shouldnt we let the metadata load from ufs happen from here and then do data load?

**Comment by lucyge2022** [comment]
is metadata sync part of the plan in this?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 9 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:158]: this is supposed to be ufs path...

partial listing is supported by the interface but is not used in the scheduler

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 10 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:207]: thanks i'll double check

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java (hunk 2 above)
Comment (elega) [reply dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:574]: how can you load the data alone without having the metadata of the file?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 3 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:150]: honestly i don't know how to capture the job exiting event. Should i close the connection in failJob() and setJobSuccess() methods?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 10 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:207]: read the code and asked other folks, it is actually the ufs path.

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 11 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:186]: okay i'll revert

File: dora/job/common/src/main/java/alluxio/scheduler/job/Task.java [hunk 13]
```diff
@@ -170,7 +170,7 @@ public boolean equals(Object obj) {
       return false;
     }
     Task<?> other = (Task<?>) obj;
-    return getPriority() == other.getPriority()
+    return getTaskId() == other.getTaskId() && getPriority() == other.getPriority()
```
Comment (elega) [inline dora/job/common/src/main/java/alluxio/scheduler/job/Task.java:173]: @lucyge2022 @jja725 please review

Comment (lucyge2022) [reply dora/job/common/src/main/java/alluxio/scheduler/job/Task.java:173]: u can remove the priority, just use task id and job is enough

File: dora/shell/src/main/java/alluxio/cli/fs/command/LoadCommand.java [hunk 14]
```diff
@@ -210,7 +219,7 @@ public int run(CommandLine cl) throws AlluxioException, IOException {
   public String getUsage() {
     return "For backward compatibility: load [--local] <path>\n"
         + "For distributed load:\n"
-        + "\tload <path> --submit [--bandwidth N] [--verify] [--partial-listing]\n"
+        + "\tload <path> --submit [--bandwidth N] [--verify] [--partial-listing] [--metadataOnly]\n"
```
Comment (jja725) [inline dora/shell/src/main/java/alluxio/cli/fs/command/LoadCommand.java:222]: I think the new convention is metadata-only

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 15]
```diff
@@ -133,88 +133,48 @@ public class DoraLoadJob extends AbstractJob<DoraLoadJob.DoraLoadTask> {
...
-    public URIStatus next() {
-      while (!mIsDone.get() && mFiles.size() < PREFETCH_SIZE * 0.2) {
-        if (advance() <= 0) {
-          break;
-        }
-      }
-      URIStatus uriStatus = mFiles.poll();
-      if (uriStatus == null) {
-        throw new NoSuchElementException();
+    try {
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:165]: do you mind extracting the mUfsStatusIterator to a class like UfsFileIterable so I can adopt this to other jobs as well?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 16]
```diff
@@ -225,28 +185,34 @@ public URIStatus next() {
...
         ++filesToLoad;
-      } catch (AlluxioException | IOException ex) {
-        if (!(ex instanceof FileDoesNotExistException)) {
+      } catch (IOException e) {
+        // The previous list or get might contain stale file metadata.
+        // For example, if a file gets removed before the worker actually loads it,
+        // the load will fail and the scheduler will retry.
+        // In such case, a FileNotFoundException might be thrown when we attempt to
+        // get the file status again, and we simply ignore that file.
+        // TODO(elega) this might result in an unlimited retry.
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:205]: this to do can be done by adding up the failure conter

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 17]
```diff
@@ -497,53 +469,58 @@ public long getDurationInSec() {
...
-              fileFailure.getFile().getAlluxioPath())) {
-            addFileFailure(fileFailure.getFile().getAlluxioPath(), fileFailure.getMessage(),
-                fileFailure.getCode());
+            doraLoadTask.getMyRunningWorker(), response.getFailuresList()));
+        for (LoadFileFailure failure : response.getFailuresList()) {
+          totalLoadedBytes -= failure.getUfsStatus().getUfsFileStatus().getContentLength();
+          if (!isHealthy() || !failure.getRetryable() || !addFilesToRetry(
+              failure.getUfsStatus().getUfsFullPath())) {
+            addFileFailure(
+                mLoadRootAlluxioUri.join(failure.getUfsStatus().getName()).toString(),
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:485]: i'm a little doubtful of the concatenation here, feel like we can just pass UfsFullPath

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 18]
```diff
@@ -225,28 +185,34 @@ public URIStatus next() {
...
     int startRetryListSize = mRetryFiles.size();
     int filesToLoad = 0;
     while (filesToLoad < RETRY_THRESHOLD
-        && i++ < startRetryListSize
-        && mRetryFiles.peek() != null) {
+        && i++ < startRetryListSize && mRetryFiles.peek() != null) {
       String path = mRetryFiles.poll();
       try {
-        URIStatus uriStatus = mFs.getStatus(new AlluxioURI(path));
+        UfsStatus uriStatus = mUfs.getStatus(path);
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:196]: actually, we can deal with FileNotFoundException when we process the response, so we can avoid extra getStatus calls which could be expensive.

File: dora/tests/src/test/java/alluxio/client/cli/fs/command/DoraLoadCommandIntegrationTest.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+    FileSystemUtils.waitForAlluxioPercentage(mFileSystem, uriA, 100);
+    FileSystemUtils.waitForAlluxioPercentage(mFileSystem, uriB, 100);
+    FileSystemUtils.waitForAlluxioPercentage(mFileSystem, uriC, 100);
+    while (!mOutput.toString().contains("SUCCEEDED")) {
+      assertEquals(0, mFsShell.run("load", "/testRoot", "--progress"));
+      Thread.sleep(1000);
+    }
+    // TODO(elega) should folders be taken into account?
+    assertTrue(mOutput.toString().contains("Files Processed: 4"));
```
Comment (jja725) [inline dora/tests/src/test/java/alluxio/client/cli/fs/command/DoraLoadCommandIntegrationTest.java:73]: let's take folder into account, but the command line output/ job progress might be confusing now since it also contains folder. Previously we don't contain folder since loadData only apply to files

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 20]
```diff
@@ -133,88 +133,48 @@ public class DoraLoadJob extends AbstractJob<DoraLoadJob.DoraLoadTask> {
...
-    @Override
-    public boolean hasNext() {
-      while (!mIsDone.get() && mFiles.size() < PREFETCH_SIZE * 0.2) {
-        if (advance() <= 0) {
-          break;
-        }
-      }
-      return mFiles.peek() != null;
-    }
+    String ufsRoot = Configuration.getString(PropertyKey.DORA_CLIENT_UFS_ROOT);
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:154]: Do you think we can move this to client side and for server-side we only use ufs full path? So for mLoadRootAlluxioPath we can just use ufs full path as well

**Review by jja725 — COMMENTED** [review]
Mostly LGTM, thanks for the contribution!

File: dora/job/common/src/main/java/alluxio/scheduler/job/Task.java (hunk 13 above)
Comment (jja725) [reply dora/job/common/src/main/java/alluxio/scheduler/job/Task.java:173]: ditto

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 15 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:165]: i don't quite understand on this... mUfsStatusIterator is obtained by calling ufs listStatusIterable(). The only thing I did in addition to calling the ufs is setting the full ufs path. I doubt if it is really meaningful to extract this 7-8 lines of code.

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 16 above)
Comment (elega) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:205]: yeah you are correct. I removed this comment. If the failure count exceeds the threshold, the job will just fail and the scheduler will stop running new tasks.

File: dora/tests/src/test/java/alluxio/client/cli/fs/command/DoraLoadCommandIntegrationTest.java (hunk 19 above)
Comment (elega) [reply dora/tests/src/test/java/alluxio/client/cli/fs/command/DoraLoadCommandIntegrationTest.java:73]: ummm let's split this message then

Comment (elega) [reply dora/tests/src/test/java/alluxio/client/cli/fs/command/DoraLoadCommandIntegrationTest.java:73]: yeah i added a Directory Processed

File: dora/tests/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java [hunk 21]
```diff
@@ -79,7 +76,6 @@ public static Collection<Object[]> data() {
           .setProperty(PropertyKey.USER_STREAMING_READER_CHUNK_SIZE_BYTES, Constants.KB)
           .setProperty(PropertyKey.WORKER_BLOCK_STORE_TYPE, "PAGE")
           .setProperty(PropertyKey.WORKER_PAGE_STORE_PAGE_SIZE, Constants.KB)
-          .setProperty(PropertyKey.WORKER_PAGE_STORE_DIRS, PAGE_STORE)
```
Comment (lucyge2022) [inline dora/tests/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java:82]: why these changes?

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 15 above)
Comment (lucyge2022) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:166]: @jja725  think such class could be developed later with elega's mUfs.listStatusIterable inside and apply the async paginating listing we talked about, and it could be replacing ufsfileiterable as well

File: dora/tests/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java (hunk 21 above)
Comment (elega) [reply dora/tests/src/test/java/alluxio/client/fs/io/DoraPositionReaderIntegrationTest.java:82]: because i updated the LocalAlluxioCluster to assigin the page store dirs automatically so there is no need to set the dir manually.

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 22]
```diff
@@ -724,16 +708,20 @@ private String getTextReport() {
           progress.append(format("\t\t%s%n", stack.toString()));
         }
       }
+      progress.append(format("\tLoad data: %s%n", mLoadData));
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:711]: this should be part of setting.

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 23]
```diff
@@ -724,16 +708,20 @@ private String getTextReport() {
...
+            mTotalByteCount == null
+                ? "" : format(" out of %s", FormatUtils.getSizeFromBytes(mTotalByteCount))));
+        if (mThroughput != null) {
+          progress.append(format("\tThroughput: %s/s%n",
+              FormatUtils.getSizeFromBytes(mThroughput)));
+        }
+        progress.append(format("\tBlock load failure rate: %.2f%%%n", mFailurePercentage));
       }
-      progress.append(format("\tBlock load failure rate: %.2f%%%n", mFailurePercentage));
       progress.append(format("\tFiles Failed: %s%n", mFailedFileCount));
```
Comment (jja725) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:725]: nit: better unify the processed/failure here since we may have directory failure. Actually, we can combine the two lines into  `files/directories processed`.

**Review by jja725 — APPROVED** [review]
LGTM, we can have another PR to fix the minor job progress wording
